### ISSUE TITLE
Bad session key get_lock design

### ISSUE THREAD

**Issue by sskaje** [bcit-ci/CodeIgniter#5382]
In both CI3 and CI4, memcached and redis session handler implementations, the session key lock designing are incorrect.

Take redis for example.

[CI4](https://github.com/bcit-ci/CodeIgniter4/blob/develop/system/Session/Handlers/RedisHandler.php), code were old, buggy but more compatible.
CI4 pseudo code inside loop is:
```
    if redis::ttl(lock_key) > 0 then 
        sleep 1s 
        continue loop

    redis::set(lock_key, ttl)
```

Yes, two concurrent requests would both bypass *ttl > 0* check, and then get the lock.
Concurrent session writing may be corrupted, but [1]concurrent session reading, or [2]one writing and one or more reading are working.

Since commit 569cf69 of [CI3](https://github.com/bcit-ci/CodeIgniter/blob/develop/system/libraries/Session/drivers/Session_redis_driver.php) , introduced from #5170 :
```
    if redis::ttl(lock_key) > 0 then 
        sleep 1s 
        continue loop

    if redis::ttl(lock_key) == key-not-exists then 
        redis::setNx(lock_key, ttl)
    else 
        redis::set(lock_key, ttl)
```
This does not help but gets more buggy.
Same as CI4 code, both requests can bypass *ttl > 0* and both hit ttl === -2 which is key not exists.
Request 1 hit redis::setNx and then success, but request 2 hit and then failed because 1 has made the key existed. Now request 2 wouldn't get session data. 

The solution is just simplified the code.

Redis:
```php

    protected function _get_lock($session_id)
    {
        // PHP 7 reuses the SessionHandler object on regeneration,
        // so we need to check here if the lock key is for the
        // correct session ID.
        if ($this->_lock_key === $this->_key_prefix.$session_id.':lock')
        {
            log_message('error', 'SESSION Redis _get_lock');
            return $this->_redis->setTimeout($this->_lock_key, 300);
        }

        // 30 attempts to obtain a lock, in case another request already has it
        $lock_key = $this->_key_prefix.$session_id.':lock';
        $attempt = 0;
        do
        {
            $result = $this->_redis->set($lock_key, time(), array('nx', 'ex'=>300));

            if (! $result)
            {
                sleep(1);
                continue;
            }

            $this->_lock_key = $lock_key;
            break;
        }
        while (++$attempt < 30);

        if ($attempt === 30) {
            // in case someone added a key without expiration
            $ttl = $this->_redis->ttl($lock_key);
            if ($ttl === -1 || $ttl > 300) {
                $this->_redis->setTimeout($lock_key, 300);
            }

            log_message('error', 'Session: Unable to obtain lock for '.$this->_key_prefix.$session_id.' after 30 attempts, aborting.');
            return FALSE;
        }

        $this->_lock = TRUE;
        return TRUE;
    }

```

Memcached
```php

	protected function _get_lock($session_id)
	{
		// PHP 7 reuses the SessionHandler object on regeneration,
		// so we need to check here if the lock key is for the
		// correct session ID.
		if ($this->_lock_key === $this->_key_prefix.$session_id.':lock')
		{
			if ( ! $this->_memcached->replace($this->_lock_key, time(), 300))
			{
				return ($this->_memcached->getResultCode() === Memcached::RES_NOTFOUND)
					? $this->_memcached->add($this->_lock_key, time(), 300)
					: FALSE;
			}

			return TRUE;
		}

		// 30 attempts to obtain a lock, in case another request already has it
		$lock_key = $this->_key_prefix.$session_id.':lock';
		$attempt = 0;
		do
		{
			if ($this->_memcached->add($lock_key, time(), 300))
			{
				sleep(1);
				continue;
			}

			$this->_lock_key = $lock_key;
			break;
		}
		while (++$attempt < 30);

		if ($attempt === 30)
		{
		    // if someone manually adds a key with a very long ttl, this session data would be not accessible by clients

			log_message('error', 'Session: Unable to obtain lock for '.$this->_key_prefix.$session_id.' after 30 attempts, aborting.');
			return FALSE;
		}

		$this->_lock = TRUE;
		return TRUE;
	}
```

**Please notice: **

1 Memcached API does not provide a command reading key's ttl, currently the value of lock key is timestamp, code inside *$attempt === 300* can be like
```php

		if ($attempt === 30)
		{
		    // if someone manually adds a key with a very long ttl, this session data would be not accessible by clients

			$lock_data = $this->_memcached->get($lock_key);
			if ($lock_data < time() - 300) {
				$this->_memcached->delete($lock_key);
			}
            
			log_message('error', 'Session: Unable to obtain lock for '.$this->_key_prefix.$session_id.' after 30 attempts, aborting.');
			return FALSE;
		}

```
I considered if a 'return $this->_get_lock($session_id);' can be added after memcached::delete(), the answer is no.
Returning false only kills one requests, but retrying to get lock might kill all concurrent requests.

2 The **300** can be changed to a constant, class constant or a session config item.

**Comment by narfbg** [comment]
This is very, very unclear.

First of all, this is the CI3 repo and talking about CI4 doesn't help, only adds confusion and makes it hard to follow.  
Secondly, I cannot find a problem statement in terms of what's the incorrect+expected end result, only hard to follow talk about TTLs and return values (which are related, of course, but are implementation details).  
And finally, you're giving short examples with pseudo-code, but then dumping entire functions that I guess we're supposed to just copy-paste, without even seeing a diff?

Please explain it better, isolated to CI3 and do start with what you expect to happen in what scenario, compared to how it works now.

**Comment by sskaje** [comment]
@narfbg This is just a discussion, CI3 and CI4's session handler are almost the same. That's why I didn't create two PRs to these two project and used pseudo code in this issue.
I will provide PRs to both projects after we make this designing clear.

First thing, compare CI3 and CI4, if they are similar, I just need to discuss this issue in either of two and the other would have same issue.

I didn't find ways compare one file, so please open links below in your desktop browser and compare between tabs. 

redis[0]: [CI4, Redis Session Handler, lockSession()](https://github.com/bcit-ci/CodeIgniter4/blob/develop/system/Session/Handlers/RedisHandler.php#L351)
redis[1]: [CI3, Redis Session Handler, _get_lock()](https://github.com/bcit-ci/CodeIgniter/blob/dec2b56fa44803e1579804fe329fd45658c6217b/system/libraries/Session/drivers/Session_redis_driver.php#L362), before #5170 
redis[2]: [CI3, Redis Session Handler, _get_lock()](https://github.com/bcit-ci/CodeIgniter/blob/3738910fce3192054cd2815d814ab261c3d3868c/system/libraries/Session/drivers/Session_redis_driver.php#L362), now with #5170 


memcached[0]: [CI4, Memcached Session Handler, lockSession()](https://github.com/bcit-ci/CodeIgniter4/blob/a6fd7a135e22084d11dfbf0c5726747d5d32ab41/system/Session/Handlers/MemcachedHandler.php#L332)
memcached[1]: [CI3, Memcached Session Handler, _get_lock()](https://github.com/bcit-ci/CodeIgniter/blob/fced25f5728ce81fe810216fcaa4ccec7523f6c9/system/libraries/Session/drivers/Session_memcached_driver.php#L323), before #5379 
memcached[2]: [CI3, Memcached Session Handler, _get_lock()](https://github.com/bcit-ci/CodeIgniter/blob/3738910fce3192054cd2815d814ab261c3d3868c/system/libraries/Session/drivers/Session_memcached_driver.php#L325), now with #5379 


Two groups, Redis and Memcached, code from each, first two links, [0] and [1], are almost the same. For CI4, latest code, for CI3, code before #5170 and #5379.

Compare two [0]s, quite similar.

Compare two [2]s, similar changes were made, both checking if key exists and then act based on key detection's result.


Till now, I think I've made this clear that we just need to discus only one implementation in CI3.


In new code, two new requests, A and B, with same session_id S, hit following code at the same time
```php
			if (($ttl = $this->_redis->ttl($lock_key)) > 0)
			{
				sleep(1);
				continue;
			}
```
Both check lock_key lock:S, and redis::ttl() return -2, telling this key is not locked.
Next, 
```php
			$result = ($ttl === -2)
				? $this->_redis->set($lock_key, time(), array('nx', 'ex' => 300))
				: $this->_redis->setex($lock_key, 300, time());
```
A and B have $ttl = -2, so they both hit setNx *$this->_redis->set($lock_key, time(), array('nx', 'ex' => 300))*
If A arrives first, B will get a *false* with **'nx'**, vice versa.

```php
			if ( ! $result)
			{
				log_message('error', 'Session: Error while trying to obtain lock for '.$this->_key_prefix.$session_id);
				return FALSE;
			}
```
So B failed to get session lock, with error like 
```
ERROR - 2018-01-17 17:06:31 --> Session: Error while trying to obtain lock for nbss:11075jfvblitou09ic772dm3e8gp11ni
ERROR - 2018-01-17 17:06:31 --> Severity: Warning --> session_start(): Failed to read session data: user (path: tcp://redis.xxx.xxx.xxx:6379?prefix=nbss:&amp;weight=1) ........../CodeIgniter/system/libraries/Session/Session.php 127
```



Back to my pseudo code, I addressed that:
1 check and act is wrong, add-if-not-exist is right choice.
2 both redis and memcached implementation facing same problem that, if someone mis-set a lock key with very long TTL or never expire, data with same session id will not be accessible by clients.


add-if-not-exist in redis:
```php
            $result = $this->_redis->set($lock_key, time(), array('nx', 'ex'=>300));
```
add-if-not-exist in memcached:
```php
            $this->_memcached->add($lock_key, time(), 300)
```

recycle bad lock key in redis:
```php
        if ($attempt === 30) {
            // in case someone added a key without expiration
            $ttl = $this->_redis->ttl($lock_key);
            if ($ttl === -1 || $ttl > 300) {
                $this->_redis->setTimeout($lock_key, 300);
            }

            log_message('error', 'Session: Unable to obtain lock for '.$this->_key_prefix.$session_id.' after 30 attempts, aborting.');
            return FALSE;
        }
```
recycle bad lock key in memcached:    no perfect solution, but a working one.
```php
		if ($attempt === 30)
		{
		    // if someone manually adds a key with a very long ttl, this session data would be not accessible by clients

			$lock_data = $this->_memcached->get($lock_key);
			if ($lock_data < time() - 300) {
				$this->_memcached->delete($lock_key);
			}
            
			log_message('error', 'Session: Unable to obtain lock for '.$this->_key_prefix.$session_id.' after 30 attempts, aborting.');
			return FALSE;
		}
```

I can't see why old session lock codes were written in that way, and also can't understand why @tianhe1986 patched in such way.

**Comment by narfbg** [comment]
> Till now, I think I've made this clear that we just need to discus only one implementation in CI3.

Discussing only a single implementation, and specifically the latest CI3 version of it is what I wanted. Yet, there's half a screen of text before this line talking about a total of 4 different code trees + 2 versions for 2 of them ...  
I wrote the basis of that algorithm, of course I know they're all similar and the same high-level solution should be applied to all. Just stick to the Redis driver, latest code in the 3.1-branch and once that's figured out, the rest is easy.

> redis[1]: CI3, Redis Session Handler, _get_lock(), before #5170
> redis[2]: CI3, Redis Session Handler, _get_lock(), now with #5170

I don't understand why you're listing these, given that you later state this as the problem:

> A and B have $ttl = -2, so they both hit setNx $this->_redis->set($lock_key, time(), array('nx', 'ex' => 300))
> If A arrives first, B will get a false with 'nx', vice versa.
>
> ...
>
> So B failed to get session lock, with error

As I understand it, you expect that B never fails unless there's a timeout and you're saying that it fails due to a race condition.
#5170 is an entirely different problem - with the exact same race condition, *both* requests would succeed, potentially leading to data corruption. The patch fixes that.

If (again) I understand correctly, you are suggesting the following patch:

```diff
@@ -335,22 +335,12 @@ class CI_Session_redis_driver extends CI_Session_driver implements SessionHandle
                $attempt = 0;
                do
                {
-                       if (($ttl = $this->_redis->ttl($lock_key)) > 0)
+                       if ( ! $this->_redis->set($lock_key, time(), array('nx', 'ex' => '300'))
                        {
                                sleep(1);
                                continue;
                        }
 
-                       $result = ($ttl === -2)
-                               ? $this->_redis->set($lock_key, time(), array('nx', 'ex' => 300))
-                               : $this->_redis->setex($lock_key, 300, time());
-
-                       if ( ! $result)
-                       {
-                               log_message('error', 'Session: Error while trying to obtain lock for '.$this->_key_prefix.$session_id);
-                               return FALSE;
-                       }
-
                        $this->_lock_key = $lock_key;
                        break;
                }
```

And that's supposed to eliminate the race condition, and make request B never fail. Is that correct?

Again, we need to focus on the main problem, so for now ignore memcached and the possible no-TTL keys - that's a different issue.

**Comment by sskaje** [comment]
@narfbg 
#5170 fix the race condition by killing one in two concurrent requests, the side effect is that one in two read-only requests is also killed.

Wait until timeout or or others release lock or reach max attempts is wiser than kill request. 
So yes, this simplified code is part of my suggestion. And long/no ttl keys problem is brought by this solution.

Another thing, I designed framework for my previous company, I introduced session read only mode for my session handler because most my APIs requires login status validation but session read only.
If session read only mode is enabled, session handler would be lock free and no write action were performed after a local stored ttl, this is quite useful for high load web sites.

I'm thinking about which can be a default session locking mode. 
Killing costs too much, it can't be a solution but a bug; 
waiting blocks other requests, it's bad if some requests is slow;
optimization like read only mode is too complex;
lock-free ?

**Comment by narfbg** [comment]
> #5170 fix the race condition by killing one in two concurrent requests, the side effect is that one in two read-only requests is also killed.

Correct.

> Wait until timeout or or others release lock or reach max attempts is wiser than kill request.  
> So yes, this simplified code is part of my suggestion.

I could agree on this.

I guess I see why you're so insistent on mentioning #5170 now, but have in mind this - previously, only setex() was used, which allowed the potential data corruption. It's way more important that that was fixed, and I guess nobody thought of your solution way back then. So there's that, no need to further talk about it.

> And long/no ttl keys problem is brought by this solution.

OK, we can discuss the >300s/no-TTL problem now.

The thing is, that doesn't happen by accident, and there's no good solution for it:

- Ignore and assume lock obtained - possible data corruption.
- Ignore and assume failure - lock can never be obtained.
- Delete/overwrite it - messing with data that we didn't create.

But either way, I don't see how that's caused by any solution to the race condition problem. And to be completely honest, since one has to intentionally set a very specific key (which includes high-entropy randomn data), whoever does that completely deserves any possible negative effects IMO.

> Another thing, I designed framework for my previous company, I introduced session read only mode for my session handler because most my APIs requires login status validation but session read only.

That's cool, but here's an off-topic note: I hope you're not talking about a REST API, because that should be stateless and thus shouldn't even know what a session is. :)

> If session read only mode is enabled, session handler would be lock free and no write action were performed after a local stored ttl, this is quite useful for high load web sites.

It can't be lock-free.

A shared/read lock could be obtained to allow simultaneous reading by multiple processes, but it's a lock nevertheless - writes have to be blocked during read times.  And there's a few problems with that when talking about CodeIgniter:

- You need to declare it read-only upfront, when the session is started (and I'd probably be the only person ever to do that instead of auto-loading the library).
- The more requests incoming, the more they would push a potential write request back into the queue (potentially forever), and the part of our userbase actually benefitting from locks are those who write AJAX-heavy apps, some with long polling - ideal conditions to create this problem. We're *emulating* locks for Redis and Memcached here, while OS-level ones might be able to avoid the issue, idk.
- You're kind of missing the point ... It's not about high-load websites - a high load is typically caused by lots of unique users, and they would have different session IDs, so they'd all have individual locks.

Not saying it's not useful, it is, but just not a right fit for CI.

> I'm thinking about which can be a default session locking mode.

The word "default" implies more configuration options - not happening, sorry. There's only going to be one way.

> Killing costs too much, it can't be a solution but a bug;

You can't classify every solution that you deem sub-optimal to be a bug ... Bugs are unintended side effects and this is not a bug. That being said, better solutions are welcome.

> waiting blocks other requests, it's bad if some requests is slow;

So make your apps fast. :)

> optimization like read only mode is too complex;

Agree, as explained above.

> lock-free ?

No. *No!* **HELL NO!**

This is a total nightmare, and we've been through it - look at the history of session-related issues with CI2. Probably 90% of them caused by concurrency and no locks.  

Introducing locking and making it mandatory was the entire point of rewriting `CI_Session` for CI 3.0 and I'd rather drop the entire library than make it lock-free. I've dedicated an entire section of the manual to warn people not to ever try dropping the locks. Not happening.

**Comment by narfbg** [comment]
I would've patched this already, but you said you wanted to discuss it before submitting a PR yourself ... Should I wait or go ahead myself?

**Comment by sskaje** [comment]
@narfbg sorry I was too busy since last week.

I just re-forked CI3 and pushed changes to my fork. 
https://github.com/sskaje/CodeIgniter/commit/c8e321c488a662ad2f5f191d4642eb4bb0dff919
There were some coding style problems, like brackets and indention, fixed in later commits.

I let lock mode configurable, because in my case I really need less lock and can't let http requests blocked.

But I'm still hesitate if lock wait would be the best default solution.

Please review. If you're ok with these, I'll create PR for both CI3 and CI4.

**Comment by narfbg** [comment]
I'm not accepting those new settings.

**Comment by guitarbien** [comment]
Hi, I got the same redis lock issue in CI3, too.
![snipaste_2018-05-25_16-59-01](https://user-images.githubusercontent.com/1098076/40535883-028647d0-603d-11e8-9d12-58c03fc5a7ce.png)
The first request hit a controller then CI returned the html back, the second is a ajax call, fetch data from an API. Both of them are read-only request.
It happened in CI 3.0.6 on my production site, but I upgrade to 3.1.8, it always happened (which was better I think so that I can debug easily).
Both requests got `-2` ttl, and one got false when `$this->_redis->set($lock_key, time(), array('nx', 'ex' => 300))`

I'm thinking about to use `Session_file_driver` in multiple server environment, and set the `sess_save_path` to a mount path (maybe symbolic link). Because the `file mode` use OS's `flock()`, it's should work better, but not confident about the speed, Redis should faster then file without emulating locking.

Any suggestion is welcome, thanks.

**Comment by kionez** [comment]
@narfbg
I'm trying to avoid the session lock issue with concurrent\parallel request (i.e.: multiple ajax calls).

If two concurrent request A and B try a setnx at the same time, the second will fail giving  the "Session: Error while trying to obtain lock for..." error. So, I'm trying to test a different solution: if the ttl of a lock key is -2 (so lock doesn't exists) try to set with expiration (setnx) but if it fails, sleep a little bit and retry from start and fail only if there is >30 attempts. I'm using usleep to reduce waiting time.

(insipred from: https://redis.io/commands/setnx#design-pattern-locking-with-codesetnxcode )

```
if (($ttl = $this->_redis->ttl($lock_key)) > 0)
{
    $sleep_random = (rand(1,5)) * 100000;
    usleep($sleep_random);
    continue;
}

if ($ttl === -2)
{
    $result = $this->_redis->set($lock_key, time(), array('nx', 'ex' => 300));
    if (!$result) {
        $sleep_random = (rand(1,5)) * 100000;
        usleep($sleep_random);
        continue;
    }
}
else {
    $result = $this->_redis->setex($lock_key, 300, time());
}
```

**Comment by narfbg** [comment]
@kionez This is a long, complicated thread that gives me a headache every time and isolated code samples are hard to analyze without comparison to our current code (i.e. there's no diff).

So, I appreciate it, but can you please just submit a pull request with your idea? I'm not saying that I'll just merge it, but even for the sake of discussion it would really make things much easier.

As a side tip though, don't use `rand()` or attempt to involve any kind of randomness in this. It doesn't help with anything.

**Comment by kionez** [comment]
Sorry, you're right.. i wrote the comment in a hurry as I tried to solve. I created a PR, hope there is something useful ;)

**Comment by erwinsetiawan** [comment]
I also got this issue when using ajax, didn’t understand why on session cache lib keep using lock since Redis didn’t support lock. I just get no problem when using $_SESSION instead ci session. Could someone solved this session library soon?

**Comment by narfbg** [comment]
So ... What are you going to do with #5585 then? Rework it or close?

**Comment by ytetsuro** [comment]
#5585 close.

**Comment by mhkhung** [comment]
We tried to upgrade to CI 3.1.10 (from much older version) and hit this problem immediately for pretty much every user: 
```
			if ( ! $this->_redis->setex($lock_key, 300, time()))
			{
				log_message('error', 'Session: Error while trying to obtain lock for '.$this->_key_prefix.$session_id);
				return FALSE;
			}
```
At least it should be changed to:
```
			if ( ! $result)
			{
				continue; // Keep on trying until that 30 attempts is up..
			}
```

**Comment by roastduck** [comment]
> > Wait until timeout or or others release lock or reach max attempts is wiser than kill request.
> > So yes, this simplified code is part of my suggestion.
> 
> I could agree on this.

It seems that this is the only thing in this thread without disputes. Is this to be implemented? I think the code pieces in the first post of this thread can resolve this, just the timeout value can be adjusted.

**Comment by narfbg** [comment]
Sorry, the only valuable part in all of this mess was this commit link:

> https://github.com/sskaje/CodeIgniter/commit/c8e321c488a662ad2f5f191d4642eb4bb0dff919

But unfortunately, @sskaje had coupled the solution with a bunch of features that I'd never accept, so it was never pulled in. And as you can see, that commit is now deleted and thus completely lost.

I'm afraid there's nothing to be salvaged from this thread, as each time someone tries to even just use it for context, it only creates more confusion. I'm closing it, so that the next time anyone wants to resolve the issue, they can start fresh, with a clear explanation.

**Comment by ryanbadger** [comment]
For anybody having issues with CI and memcached sessions, I found that this:

$this->_memcached->touch($key, $this->_config['expiration'])
OR ($this->_memcached->getResultCode() === Memcached::RES_NOTFOUND && $this->_memcached->set($key, $session_data, $this->_config['expiration']))

was causing about a 5s delay, even though the caching actually worked fine.

I commented out the above and now my site loads in milliseconds rather than seconds.

I'm sure there's a very good reason for that line existing, so please do share if there's a way of keeping the functionality without adding 5s load time.
